pkg/types/config/context.go

```diff
@@ -246,6 +230,18 @@ type UserContext struct {
 	KindNamespaces map[schema.GroupVersionKind]string
 }
 
+func (c *ManagementContext) WithAgent(userAgent string) ManagementContext {
```

Comment:
One downfall of returning a "new" `ManagementContext` here is our shallow copy causes some nuance behavior. `mgmtCopy` and `c` are now oddly coupled because some of their member variables, like `ClientGetter` and `UnvsersionedClient` point to the same object while others like `Management` and `RBAC` are now completely different. So now the caller has to know the implementation details of WithAgent to determine if changes to the context will affect others or not.

Reply:
@rmweir 
>This is by design. We don't want to deep copy values where it isn't necessary. For the interface we may be returning a new struct, but we are passing the factory which makes up a majority of the new struct along with its CacheFactory. The Client structure itself, the ShareClientFactory, and the ShareClientFactory's userAgent field are the only new values asides from pointers referencing existing structures.

Okay, including in the comments that the returned object is a shallow copy would help prevent a future caller from mutating the returned context without realizing it is used by others.